fennel/test_lib/query_engine.py

```diff
@@ -0,0 +1,392 @@
+from collections import defaultdict
+from datetime import datetime
+from functools import partial
+from typing import Dict, List, Union, Optional, Any, Tuple
+
+import numpy as np
+import pandas as pd
+from frozendict import frozendict
+
+import fennel.datasets.datasets
+from fennel.featuresets import Extractor, Feature, Featureset, is_valid_feature
+from fennel.gen.featureset_pb2 import (
+    ExtractorType as ProtoExtractorType,
+)
+from fennel.gen.schema_pb2 import Field, DSSchema, Schema
+from fennel.lib.schema import data_schema_check, get_datatype
+from fennel.test_lib.branch import Entities
+from fennel.test_lib.data_engine import DataEngine
+from fennel.test_lib.test_utils import cast_col_to_dtype
+import fennel.gen.schema_pb2 as schema_proto
+
+
+class QueryEngine:
+    """
+    Query engine is a stateless class that handles all things related to lookups and extract given
```

Comment:
The Query Engine is a stateless class responsible for managing all aspects of lookups, including the execution of specified extractors and interaction with the data engine.

Reply:
changed